traversal/procedure/ProcedureVertex.java

```diff
@@ -88,27 +90,38 @@ public int order() {
         return order;
     }
 
-    public List<ProcedureEdge<?, ?>> orderedOuts() {
-        if (orderedOuts == null) {
-            orderedOuts = outs().stream().sorted(Comparator.comparingInt(e -> e.to().order())).collect(toList());
+    void setOrder(int order) {
+        this.order = order;
+    }
+
+    public Set<ProcedureVertex<?, ?>> dependents() {
+        if (dependents == null) {
+            Set<ProcedureVertex<?, ?>> vertices = new HashSet<>();
+            outs().forEach(e -> vertices.add(e.to()));
+            dependents = vertices;
         }
```

Comment:
this is because this `Procedure` classes are shared across threads. To avoid the possibility of race conditions and not require locking, we first in each thread collect to a set and then do an atomic assignment operation (assignment is always atomic in Java), where one writer will win. 
Technically because set addition is idempotent we could avoid doing this if the set we use also supports concurrent writers - however, using a `ConcurrentSet` will also incur read time overheads we don't want to pay.